app/src/main/java/pl/ppiwd/exerciseanalyst/activities/TrainingDataActivity.java

```diff
@@ -0,0 +1,254 @@
+    private ServerConnection serverConnection;
+    private static final String[] activities = new String[]{
+            "Walking",
+            "Frontal elevation of arms",
+            "Jump front & back",
+            "Squats",
+            "Push ups"
+    };
+    private Intent deviceService;
+
+    private void initialize() {
+        this.timerHandler = new Handler();
+        this.timerTextView = findViewById(R.id.timerTextView);
+        this.timer = new Timer(timerTextView, timerHandler);
+        this.startStopButton = findViewById(R.id.startSessionButton);
+        this.activitySpinner = getSpinner();
+        this.repetitionsPicker = getPicker();
+        this.startStopButton.setOnClickListener(view -> handleSession());
+
+        if (getIntent().getStringExtra(Constants.TRAINING_TYPE).equals(Constants.TRAINING_REGULAR)) {
+            findViewById(R.id.activityTextView).setVisibility(View.INVISIBLE);
+            findViewById(R.id.repetitionsTextView).setVisibility(View.INVISIBLE);
+            activitySpinner.setVisibility(View.INVISIBLE);
+            repetitionsPicker.setVisibility(View.INVISIBLE);
+        } else {
+            findViewById(R.id.backgroundImage).setVisibility(View.INVISIBLE);
+        }
+
+        this.deviceService = new Intent(this, MetaMotionService.class);
+        this.serverConnection = getServerConnection();
+        this.deviceConnectionChecker = new DeviceConnectionChecker(this, metaMotionServiceConnection);
+    }
+
+    private ServerConnection getServerConnection() {
+        MeasurementsDao measurements = MeasurementsDatabase.getInstance(getApplicationContext()).measurementsDao();
+        return new ServerConnection(measurements, this);
+    }
+
+    private Spinner getSpinner() {
+        ArrayAdapter<String> adapter = new ArrayAdapter<>(TrainingDataActivity.this, android.R.layout.simple_spinner_item, activities);
+        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+        Spinner spinner = (Spinner) findViewById(R.id.activitySpinner);
+        spinner.setAdapter(adapter);
+        return spinner;
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_training);
+        initialize();
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        setResult(RESULT_CANCELED, null);
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        if (deviceConnectionChecker.isConnected()) {
+            bindToMetaMotionService();
+        }
+    }
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+        finishSession();
+        deviceConnectionChecker.unregister();
+        serverConnection.cancelRequest();
+        Toast.makeText(this, "Unfinished trainings dropped", Toast.LENGTH_SHORT).show();
+    }
```

Comment:
You're right, there were cases that I didn't test before. I know that it's a bit late for such changes with binding, but I think that since the whole session happens in a single view/Action and not in two of them as was previously, there is no need to use bindings instead of start/finish Service.

I've been testing it for a while and my solution works fine with: rotating screen, keyboard showing up, locking the screen, and minimizing application.